File: src/main/java/ninja/hudy/infosmog/model/VoivodeshipRegion.java
package ninja.hudy.infosmog.model;

import javax.persistence.DiscriminatorValue;
import javax.persistence.Entity;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@DiscriminatorValue(RegionType.Values.VOIVODESHIP)
public class VoivodeshipRegion extends Region {

    public Set<CountyRegion> getCouties() {
        return getNestedRegions().stream()
                .filter(region -> RegionType.COUNTY
                        .toString()
                        .equals(region.getType()))
                .map(region -> (CountyRegion) region)
                .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/pl/org/smok/infosmog/model/VoivodeshipRegion.kt
package pl.org.smok.infosmog.model

import javax.persistence.DiscriminatorValue
import javax.persistence.Entity
import java.util.stream.Collectors

@Entity
@DiscriminatorValue(RegionType.Values.VOIVODESHIP)
class VoivodeshipRegion : Region() {

    val couties: Set<CountyRegion>
        get() = nestedRegions.stream()
                .filter { region ->
                    RegionType.COUNTY
                            .toString() == region.type
                }
                .map { region -> region as CountyRegion }
                .collect<Set<CountyRegion>, Any>(Collectors.toSet<CountyRegion>())
}
